internal/agent/device/spec/notifier.go

```diff
@@ -0,0 +1,124 @@
+package spec
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/agent/client"
+	"github.com/flightctl/flightctl/internal/agent/device/errors"
+	"github.com/flightctl/flightctl/pkg/log"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type Notifier interface {
+	PollAndNotify(ctx context.Context)
+	Subscribe() chan *v1alpha1.Device
+	SetClient(client.Management)
+}
+
+type notifier struct {
+	managementClient client.Management
+	deviceName       string
+	subscribers      []chan *v1alpha1.Device
+	lastKnownVersion string
+	log              *log.PrefixLogger
+	backoff          wait.Backoff
+	mu               sync.Mutex
+}
+
+func NewNotifier(deviceName string,
+	backoff wait.Backoff,
+	log *log.PrefixLogger) Notifier {
+	return &notifier{
+		deviceName: deviceName,
+		backoff:    backoff,
+		log:        log,
+	}
+}
+
+func (n *notifier) getRenderedFromManagementAPIWithRetry(
+	ctx context.Context,
+	renderedVersion string,
+	rendered *v1alpha1.Device,
+) (bool, error) {
+	params := &v1alpha1.GetRenderedDeviceParams{}
+	if renderedVersion != "" {
+		params.KnownRenderedVersion = &renderedVersion
+	}
+
+	resp, statusCode, err := n.managementClient.GetRenderedDevice(ctx, n.deviceName, params)
+	if err != nil {
+		return false, fmt.Errorf("%w: %w", errors.ErrGettingDeviceSpec, err)
+	}
+
+	switch statusCode {
+	case http.StatusOK:
+		if resp == nil {
+			// 200 OK but response is nil
+			return false, errors.ErrNilResponse
+		}
+		*rendered = *resp
+		return true, nil
+
+	case http.StatusNoContent, http.StatusConflict:
+		// instead of treating it as an error indicate that no new content is available
+		return true, errors.ErrNoContent
+
+	default:
+		// unexpected status codes
+		return false, fmt.Errorf("%w: unexpected status code %d", errors.ErrGettingDeviceSpec, statusCode)
+	}
+}
+
+func (n *notifier) PollAndNotify(ctx context.Context) {
+	newDesired := &v1alpha1.Device{}
+
+	startTime := time.Now()
+	err := wait.ExponentialBackoff(n.backoff, func() (bool, error) {
+		return n.getRenderedFromManagementAPIWithRetry(ctx, n.lastKnownVersion, newDesired)
+	})
+
+	// log slow calls
+	duration := time.Since(startTime)
+	if duration > time.Minute {
+		n.log.Debugf("Dialing management API took: %v", duration)
+	}
+	if err != nil {
+		if errors.Is(err, errors.ErrNoContent) || errors.IsTimeoutError(err) {
+			n.log.Debug("No new template version from management service")
+			return
+		}
+		n.log.Errorf("Received non-retryable error from management service: %v", err)
+		return
+	}
+
+	n.lastKnownVersion = newDesired.Version()
+
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	// notify all subscribers of the new device spec
+	for _, ch := range n.subscribers {
```

Comment:
does this channel need to be closed?

Reply:
It is an slice of channels.  Not a single channel